posts/2021-07-29-Rust-1.54.0.md

```diff
@@ -0,0 +1,113 @@
+
+### Attributes can call macros!()
+
+Rust 1.54 supports invoking function-like macros inside attributes. One notable use case for this is including documentation from other files into Rust doc comments. For example, if your project's README represents a good documentation comment, you can use `include_str!` to directly incorporate the contents. Previously, various workarounds allowed similar functionality, but from 1.54 this is much more ergonomic.
+
+
+```rust=
+#![doc = include_str!("README.md")]
+```
+
+Macros can be nested inside the attribute as well, for example to include content generated by a build script:
+
+```rust
+#[path = concat!(env!("OUT_DIR"), "/generated.rs")]
+mod generated;
+```
+
+Read [here](https://github.com/rust-lang/rust/pull/83366) for more details.
+
+### wasm32 intrinsics stabilized
+
+A number of intrinsics for the wasm32 platform have been stabilized, which gives access to the SIMD instructions in WebAssembly.
+
+Notably, unlike the previously stabilized `x86` and `x86_64` intrinsics, these do not have a safety requirement to only be called when the appropriate target feature is enabled. This is because WebAssembly was written from the start to validate code safely before executing it, so instructions are guaranteed to be decoded correctly (or not at all).
+
+This means that we can expose some of the intrinsics as entirely safe functions, for example [`v128_bitselect`](https://doc.rust-lang.org/beta/core/arch/wasm32/fn.v128_bitselect.html). However, there are still some intrinsics which are unsafe because they use raw pointers, such as [`v128_load`](https://doc.rust-lang.org/beta/core/arch/wasm32/fn.v128_load.html).
+
+### Incremental Compilation is re-enabled by default
+
+Incremental compilation has been re-enabled by default in this release, after it being disabled by default in 1.52.1.
+
+In Rust 1.52, additional validation was added when loading incremental compilation data from the on-disk cache.
+This resulted in a number of pre-existing potential soundness issues being uncovered as the validation changed these silent bugs into internal compiler errors (ICEs).
+In response, the Compiler Team decided to disable incremental compilation in the 1.52.1 patch, allowing users to avoid encountering the ICEs and the underlying unsoundness, at the expense of longer compile times. [^1]
+
+Since then, we've conducted a [series of retrospectives][retros] and contributors have been hard at work resolving the reported issues, with some fixes landing in 1.53 and the majority landing in this release. [^2]
+
+There are currently still two known issues which can result in an ICE.
+Due to the lack of automated crash reporting, we can't be certain of the full extent of impact of the outstanding issues. However, based on the feedback we received from users affected by the 1.52 release, we believe the remaining issues to be rare in practice.
+
+Therefore, incremental compilation has been re-enabled in this release!
+
```

Comment:
````suggestion
The Rust compiler has also been updated to include a more detailed error message with instructions to help the user work around the issue if it were to occur:

```
   Compiling example v0.1.0 (/path/to/example)
error: internal compiler error: encountered incremental compilation error with query(hash)
  |
  = help: This is a known issue with the compiler. Run `cargo clean -p example` or `cargo clean` to allow your project to compile
  = note: Please follow the instructions below to create a bug report with the provided information
  = note: See <https://github.com/rust-lang/rust/issues/84970> for more information

thread 'rustc' panicked at 'Found unstable fingerprints for query(hash): QueryResult', /rustc/compiler/rustc_query_system/src/query/plumbing.rs:619:9
```

````

Reply:
Should the note take the further step and actively emphasize that we really do want people to follow through and file bugs when they see this message, since, as alluded to above, that is our main avenue for feedback here?

I'm torn about this, because I don't want to turn the release blog post into a post just about the fingerprint bug. After all, our hope here is that this really will be rare now...

Reply:
I had a similar thought while writing this part and decided against it because the error message says to report the issue. I'll go ahead and add that and we can see if we like that better.

Reply:
Should we make it longer than that? I think short and to the point is better here and we want to get reports of all ICEs, not just this one.

Reply:
I'm leaning towards not including the error message: it's great that the diagnostics improved for this, but focusing on the error message seems... wrong for the release announcement :)

Reply:
I think I agree with @pietroalbini here. The error already has a call to action... Anyone who should care will see it and hopefully that is enough.